### PR TITLE
Add a Node.PY subsystem.

### PR DESCRIPTION
**Pull request by dmoody256**
This issue was originally created at: 2001-07-19 22:00:00.
This issue was reported by: `stevenknight`.
stevenknight said at 2001-07-19 22:00:00
>Tracks dependencies on in-core Python values.  Dictionaries should be equivalent to directories, strings should be "files."  This will allow us to write build-engine dependency tests more efficiently, without having to set up real files.

issues@scons said at 2001-07-19 22:00:00
>Converted from SourceForge task item 34793

### CODE REVIEW

**Comment by dmoody256** [comment]
added fix for flattening source list first seen here:
https://pairlist4.pair.net/pipermail/scons-users/2017-November/006469.html

**Comment by mwichmann** [comment]
Looking reasonable to me.  

This is just nit-picking, and I would certainly not oppose merging without it, but I'm unfond of "magic numbers" in slices, so I'd suggest a change in the three places that check for a manifest from this style:

            if contents[:16] != "Manifest-Version":

to this style:

            if not contents.startswith("Manifest-Version"):

**Comment by dmoody256** [comment]
Added a fix and test for https://pairlist4.pair.net/pipermail/scons-users/2017-November/006515.html

**Comment by dmoody256** [comment]
re-based this PR to remove the travis test script so as to limit scope and focus of the PR

File: test/Java/JAR.py [hunk 1]
```diff
@@ -134,7 +134,7 @@
 foo.Java(target = 'classes', source = 'com/sub/foo')
 bar.Java(target = 'classes', source = 'com/sub/bar')
 foo.Jar(target = 'foo', source = 'classes/com/sub/foo')
-bar.Jar(target = 'bar', source = 'classes/com/sub/bar')
+bar.Jar(target = 'bar', source = Dir('classes/com/sub/bar'))
```
Comment (bdbaddog) [inline test/Java/JAR.py:137]: Does this mean if source specify a directory going forward the value will have to be a Dir() instead of a plain string?

Comment (dmoody256) [reply test/Java/JAR.py:137]: no this test is testing both ways in this case. We didn't have a test for both cases before, all our tests do plain string.

Comment (bdbaddog) [reply test/Java/JAR.py:137]: so the Jar above with classes/com/sub/foo is also a directory?

Comment (dmoody256) [reply test/Java/JAR.py:137]: Yeah the Java builder will make a directory 'classes' which contains the correct dir structure based off the package for each java file, then compile the java files to class in that classes directory.

**Pull request merged**